knox.go

```diff
@@ -145,6 +166,79 @@ func (s PrincipalType) MarshalJSON() ([]byte, error) {
 	}
 }
 
+// IsValidPrincipal verifies that the given id string matches our expectations
+// for what a principal should look like given the principal type. For example,
+// a service principal should be a valid SPIFFE ID.
+func (s PrincipalType) IsValidPrincipal(id string, extraValidators []PrincipalValidator) error {
+	// Empty principals are invalid. Note we must deny addition of ACLs with
+	// empty MachinePrefix to prevent abuse. An empty prefix would grant *every*
+	// client access, which we don't support in Knox (if it's not a secret it
+	// doesn't belong in Knox). For other types of principals, an empty string is
+	// also invalid (though would not broadly open up access).
+	if id == "" {
+		return ErrACLEmptyPrincipal
+	}
+
+	// Apply additional validation for service and service prefix principals.
+	switch s {
+	case Service, ServicePrefix:
+		// For Service principals and prefixes, verify that id looks like a valid SPIFFE ID.
+		parsed, err := url.Parse(id)
+		if err != nil || parsed.Scheme != spiffeScheme || parsed.Host == "" {
+			return ErrACLInvalidServicePrefixURL
+		}
+
+		// Prefixes must also end in "/" as we don't allow prefixes to match inside
+		// of a path component, to avoid perfixes like "foo/bar" also matching
+		// things like "foo/barbaz" even though it should only match "foo/bar/baz".
+		endsWithSlash := strings.HasSuffix(id, "/")
+		if s == ServicePrefix && !endsWithSlash {
+			return ErrACLInvalidServicePrefixNoSlash
+		}
+	}
+
+	for _, extraValidator := range extraValidators {
+		err := extraValidator(s, id)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ServicePrefixPathComponentsValidator is an extra validator that can be applied
+// to ensure that service prefixes have a certain minimum length, e.g. to prevent
+// allow the prefix to be a full domain.
+func ServicePrefixPathComponentsValidator(minPathComponents int) PrincipalValidator {
+	return func(pt PrincipalType, id string) error {
+		if pt != ServicePrefix {
+			return nil
+		}
+
+		parsed, err := url.Parse(id)
+		if err != nil || parsed.Scheme != spiffeScheme || parsed.Host == "" {
+			return ErrACLInvalidServicePrefixURL
+		}
+
+		// Trim leading and trailing slashes, and count remaining slashes to determine
+		// the number of components in the namespace path.
+		trimmed := strings.Trim(parsed.Path, "/")
```

Comment:
https://golang.org/pkg/path/#pkg-overview

might have better support for doing what you need.

Reply:
Didn't see anything useful in there, I could split into file/dir in a loop I suppose but just counting slashes seems easier.